File: code/packages/python/state-machine/src/state_machine/dfa.py
"""Deterministic Finite Automaton (DFA) — the workhorse of state machines.

=== What is a DFA? ===

A DFA is the simplest kind of state machine. It has a fixed set of states,
reads input symbols one at a time, and follows exactly one transition for
each (state, input) pair. There is no ambiguity, no guessing, no backtracking.

Formally, a DFA is a 5-tuple (Q, Σ, δ, q₀, F):

    Q  = a finite set of states
    Σ  = a finite set of input symbols (the "alphabet")
    δ  = a transition function: Q × Σ → Q
    q₀ = the initial state (q₀ ∈ Q)
    F  = a set of accepting/final states (F ⊆ Q)

=== Why "deterministic"? ===

"Deterministic" means there is exactly ONE next state for every (state, input)
combination. Given the same starting state and the same input sequence, a DFA
always follows the same path and reaches the same final state. This makes DFAs
predictable, efficient, and easy to implement in hardware — which is why they
appear everywhere from CPU branch predictors to network protocol handlers.

=== Example: a turnstile ===

A turnstile at a subway station has two states: locked and unlocked.
Insert a coin → it unlocks. Push the arm → it locks.

    States:      {locked, unlocked}
    Alphabet:    {coin, push}
    Transitions: (locked, coin) → unlocked
                 (locked, push) → locked
                 (unlocked, coin) → unlocked
                 (unlocked, push) → locked
    Initial:     locked
    Accepting:   {unlocked}

This DFA answers the question: "after this sequence of coin/push events,
is the turnstile unlocked?"

=== Connection to existing code ===

The 2-bit branch predictor in the branch-predictor package (D02) is a DFA:

    States:      {SNT, WNT, WT, ST}  (strongly/weakly not-taken/taken)
    Alphabet:    {taken, not_taken}
    Transitions: defined by the saturating counter logic
    Initial:     WNT
    Accepting:   {WT, ST}  (states that predict "taken")

The CPU pipeline (D04) is a linear DFA: FETCH → DECODE → EXECUTE → repeat.
The lexer's character dispatch is an implicit DFA where character classes
determine transitions.
"""

from __future__ import annotations

from state_machine.types import Action, TransitionRecord


class DFA:
    """Deterministic Finite Automaton.

    A DFA is always in exactly one state. Each input causes exactly one
    transition. If no transition is defined for the current (state, input)
    pair, processing that input raises a ValueError.

    All transitions are traced via TransitionRecord objects, providing
    complete execution history for debugging and visualization.

    Example:
        >>> turnstile = DFA(
        ...     states={"locked", "unlocked"},
        ...     alphabet={"coin", "push"},
        ...     transitions={
        ...         ("locked", "coin"): "unlocked",
        ...         ("locked", "push"): "locked",
        ...         ("unlocked", "coin"): "unlocked",
        ...         ("unlocked", "push"): "locked",
        ...     },
        ...     initial="locked",
        ...     accepting={"unlocked"},
        ... )
        >>> turnstile.process("coin")
        'unlocked'
        >>> turnstile.accepts(["coin", "push", "coin"])
        True
    """

    # === Construction ===
    #
    # We validate all inputs eagerly in __init__ so that errors are caught
    # at definition time, not at runtime when the machine processes its
    # first input. This is the "fail fast" principle.

    def __init__(
        self,
        states: set[str],
        alphabet: set[str],
        transitions: dict[tuple[str, str], str],
        initial: str,
        accepting: set[str],
        actions: dict[tuple[str, str], Action] | None = None,
    ) -> None:
        """Create a new DFA.

        Args:
            states: The finite set of states. Must be non-empty.
            alphabet: The finite set of input symbols. Must be non-empty.
            transitions: Mapping from (state, event) to target state.
                Every target must be in `states`. Not every (state, event)
                pair needs a transition — missing transitions cause errors
                at processing time.
            initial: The starting state. Must be in `states`.
            accepting: The set of accepting/final states. Must be a subset
                of `states`. Can be empty (the machine never accepts).
            actions: Optional mapping from (state, event) to a callback
                function that fires when that transition occurs.

        Raises:
            ValueError: If any validation check fails.
        """
        # --- Validate states ---
        if not states:
            raise ValueError("States set must be non-empty")

        # --- Validate initial state ---
        if initial not in states:
            raise ValueError(
                f"Initial state '{initial}' is not in the states set "
                f"{sorted(states)}"
            )

        # --- Validate accepting states ---
        invalid_accepting = accepting - states
        if invalid_accepting:
            raise ValueError(
                f"Accepting states {sorted(invalid_accepting)} are not in "
                f"the states set {sorted(states)}"
            )

        # --- Validate transitions ---
        #
        # Every transition must go FROM a known state ON a known event
        # TO a known state. We check all three.
        for (source, event), target in transitions.items():
            if source not in states:
                raise ValueError(
                    f"Transition source '{source}' is not in the states set"
                )
            if event not in alphabet:
                raise ValueError(
                    f"Transition event '{event}' is not in the alphabet "
                    f"{sorted(alphabet)}"
                )
            if target not in states:
                raise ValueError(
                    f"Transition target '{target}' (from "
                    f"({source}, {event})) is not in the states set"
                )

        # --- Validate actions ---
        if actions:
            for source, event in actions:
                if (source, event) not in transitions:
                    raise ValueError(
                        f"Action defined for ({source}, {event}) but no "
                        f"transition exists for that pair"
                    )

        # --- Store the 5-tuple + extras ---
        self._states: frozenset[str] = frozenset(states)
        self._alphabet: frozenset[str] = frozenset(alphabet)
        self._transitions: dict[tuple[str, str], str] = dict(transitions)
        self._initial: str = initial
        self._accepting: frozenset[str] = frozenset(accepting)
        self._actions: dict[tuple[str, str], Action] = dict(actions or {})

        # --- Mutable execution state ---
        self._current: str = initial
        self._trace: list[TransitionRecord] = []

    # === Properties ===

    @property
    def states(self) -> frozenset[str]:
        """The finite set of states."""
        return self._states

    @property
    def alphabet(self) -> frozenset[str]:
        """The finite set of input symbols."""
        return self._alphabet

    @property
    def transitions(self) -> dict[tuple[str, str], str]:
        """The transition function as a dictionary."""
        return dict(self._transitions)

    @property
    def initial(self) -> str:
        """The initial state."""
        return self._initial

    @property
    def accepting(self) -> frozenset[str]:
        """The set of accepting/final states."""
        return self._accepting

    @property
    def current_state(self) -> str:
        """The state the machine is currently in."""
        return self._current

    @property
    def trace(self) -> list[TransitionRecord]:
        """The execution trace — a list of all transitions taken so far."""
        return list(self._trace)

    # === Processing ===

    def process(self, event: str) -> str:
        """Process a single input event and return the new state.

        Looks up the transition for (current_state, event), moves to the
        target state, executes the action (if defined), logs a
        TransitionRecord, and returns the new current state.

        Args:
            event: An input symbol from the alphabet.

        Returns:
            The new current state after the transition.

        Raises:
            ValueError: If the event is not in the alphabet, or if no
                transition is defined for (current_state, event).

        Example:
            >>> m = DFA(states={"a","b"}, alphabet={"x"},
            ...         transitions={("a","x"):"b", ("b","x"):"a"},
            ...         initial="a", accepting={"b"})
            >>> m.process("x")
            'b'
            >>> m.current_state
            'b'
        """
        # Validate the event
        if event not in self._alphabet:
            raise ValueError(
                f"Event '{event}' is not in the alphabet "
                f"{sorted(self._alphabet)}"
            )

        # Look up the transition
        key = (self._current, event)
        if key not in self._transitions:
            raise ValueError(
                f"No transition defined for (state='{self._current}', "
                f"event='{event}')"
            )

        target = self._transitions[key]

        # Execute the action if one exists
        action_name: str | None = None
        if key in self._actions:
            action = self._actions[key]
            action(self._current, event, target)
            action_name = getattr(action, "__name__", str(action))

        # Log the transition
        record = TransitionRecord(
            source=self._current,
            event=event,
            target=target,
            action_name=action_name,
        )
        self._trace.append(record)

        # Move to the new state
        self._current = target
        return target

    def process_sequence(self, events: list[str]) -> list[TransitionRecord]:
        """Process a sequence of inputs and return the trace.

        Each input is processed in order. The full trace of transitions
        is returned. The machine's state is updated after each input.

        Args:
            events: A list of input symbols.

        Returns:
            A list of TransitionRecord objects, one per input.

        Example:
            >>> m = DFA(states={"a","b"}, alphabet={"x"},
            ...         transitions={("a","x"):"b", ("b","x"):"a"},
            ...         initial="a", accepting={"b"})
            >>> trace = m.process_sequence(["x", "x", "x"])
            >>> [(t.source, t.target) for t in trace]
            [('a', 'b'), ('b', 'a'), ('a', 'b')]
        """
        trace_start = len(self._trace)
        for event in events:
            self.process(event)
        return self._trace[trace_start:]

    def accepts(self, events: list[str]) -> bool:
        """Check if the machine accepts the input sequence.

        Processes the entire sequence and returns True if the machine
        ends in an accepting state.

        IMPORTANT: This method does NOT modify the machine's current state
        or trace. It runs on a fresh copy starting from the initial state.

        Args:
            events: A list of input symbols.

        Returns:
            True if the machine ends in an accepting state after
            processing all inputs, False otherwise.

        Example:
            >>> turnstile = DFA(
            ...     states={"locked", "unlocked"},
            ...     alphabet={"coin", "push"},
            ...     transitions={
            ...         ("locked", "coin"): "unlocked",
            ...         ("locked", "push"): "locked",
            ...         ("unlocked", "coin"): "unlocked",
            ...         ("unlocked", "push"): "locked",
            ...     },
            ...     initial="locked",
            ...     accepting={"unlocked"},
            ... )
            >>> turnstile.accepts(["coin"])
            True
            >>> turnstile.accepts(["coin", "push"])
            False
            >>> turnstile.accepts([])
            False
        """
        # Run on a copy so we don't modify this machine's state
        state = self._initial
        for event in events:
            if event not in self._alphabet:
                raise ValueError(
                    f"Event '{event}' is not in the alphabet "
                    f"{sorted(self._alphabet)}"
                )
            key = (state, event)
            if key not in self._transitions:
                return False
            state = self._transitions[key]
        return state in self._accepting

    def reset(self) -> None:
        """Reset the machine to its initial state and clear the trace.

        After reset, the machine is in the same state as when it was
        first constructed — as if no inputs had ever been processed.
        """
        self._current = self._initial
        self._trace = []

    # === Introspection ===
    #
    # These methods analyze the structure of the DFA itself, not its
    # execution. They answer questions like "is the DFA well-formed?"
    # and "which states can actually be reached?"

    def reachable_states(self) -> frozenset[str]:
        """Return the set of states reachable from the initial state.

        Uses breadth-first search over the transition graph. A state is
        reachable if there exists any sequence of inputs that leads from
        the initial state to that state.

        States that are defined but not reachable are "dead weight" —
        they can never be entered and can be safely removed during
        minimization.

        Returns:
            A frozenset of reachable state names.
        """
        # BFS from the initial state
        visited: set[str] = set()
        queue: list[str] = [self._initial]

        while queue:
            state = queue.pop(0)
            if state in visited:
                continue
            visited.add(state)

            # Find all states reachable from this one via any input
            for (source, _event), target in self._transitions.items():
                if source == state and target not in visited:
                    queue.append(target)

        return frozenset(visited)

    def is_complete(self) -> bool:
        """Check if a transition is defined for every (state, input) pair.

        A complete DFA never gets "stuck" — every state handles every
        input. Textbook DFAs are usually complete (missing transitions
        go to an explicit "dead" or "trap" state). Practical DFAs often
        omit transitions to save space, treating missing transitions as
        errors.

        Returns:
            True if every (state, event) pair has a defined transition.
        """
        for state in self._states:
            for event in self._alphabet:
                if (state, event) not in self._transitions:
                    return False
        return True

    def validate(self) -> list[str]:
        """Check for common issues and return a list of warnings.

        Checks performed:
        - Unreachable states (defined but never entered)
        - Missing transitions (incomplete DFA)
        - Accepting states that are unreachable

        Returns:
            A list of warning messages. Empty if no issues found.
        """
        warnings: list[str] = []

        # Check for unreachable states
        reachable = self.reachable_states()
        unreachable = self._states - reachable
        if unreachable:
            warnings.append(
                f"Unreachable states: {sorted(unreachable)}"
            )

        # Check for unreachable accepting states
        unreachable_accepting = self._accepting - reachable
        if unreachable_accepting:
            warnings.append(
                f"Unreachable accepting states: "
                f"{sorted(unreachable_accepting)}"
            )

        # Check for missing transitions
        missing: list[str] = []
        for state in sorted(self._states):
            for event in sorted(self._alphabet):
                if (state, event) not in self._transitions:
                    missing.append(f"({state}, {event})")
        if missing:
            warnings.append(
                f"Missing transitions: {', '.join(missing)}"
            )

        return warnings

    # === Visualization ===

    def to_dot(self) -> str:
        """Return a Graphviz DOT representation of this DFA.

        Accepting states are drawn as double circles (doublecircle shape).
        The initial state has an invisible node pointing to it (the
        standard convention for marking the start state in automata
        diagrams).

        The output can be rendered with:
            dot -Tpng machine.dot -o machine.png

        Returns:
            A string in DOT format.
        """
        lines: list[str] = []
        lines.append("digraph DFA {")
        lines.append("    rankdir=LR;")
        lines.append("")

        # Invisible start node pointing to initial state
        lines.append('    __start [shape=point, width=0.2];')
        lines.append(f'    __start -> "{self._initial}";')
        lines.append("")

        # Accepting states get double circles
        for state in sorted(self._states):
            shape = (
                "doublecircle" if state in self._accepting else "circle"
            )
            lines.append(f'    "{state}" [shape={shape}];')
        lines.append("")

        # Transitions as labeled edges
        # Group transitions with same source and target to combine labels
        edge_labels: dict[tuple[str, str], list[str]] = {}
        for (source, event), target in sorted(self._transitions.items()):
            key = (source, target)
            if key not in edge_labels:
                edge_labels[key] = []
            edge_labels[key].append(event)

        for (source, target), labels in sorted(edge_labels.items()):
            label = ", ".join(sorted(labels))
            lines.append(f'    "{source}" -> "{target}" [label="{label}"];')

        lines.append("}")
        return "\n".join(lines)

    def to_ascii(self) -> str:
        """Return an ASCII transition table.

        Example output for the turnstile:

                  │ coin     │ push
        ─────────┼──────────┼──────────
        locked   │ unlocked │ locked
        unlocked │ unlocked │ locked

        Accepting states are marked with (*). The initial state is
        marked with (>).

        Returns:
            A formatted ASCII table string.
        """
        sorted_events = sorted(self._alphabet)
        sorted_states = sorted(self._states)

        # Calculate column widths
        state_width = max(len(s) + 4 for s in sorted_states)  # +4 for markers
        event_width = max(
            max((len(e) for e in sorted_events), default=0),
            max(
                (
                    len(self._transitions.get((s, e), "—"))
                    for s in sorted_states
                    for e in sorted_events
                ),
                default=0,
            ),
        )
        event_width = max(event_width, 5)  # minimum column width

        # Header row
        header = " " * state_width + "│"
        for event in sorted_events:
            header += f" {event:<{event_width}} │"
        lines = [header]

        # Separator
        sep = "─" * state_width + "┼"
        for _ in sorted_events:
            sep += "─" * (event_width + 2) + "┼"
        sep = sep[:-1]  # remove trailing ┼
        lines.append(sep)

        # Data rows
        for state in sorted_states:
            # Mark initial and accepting states
            markers = ""
            if state == self._initial:
                markers += ">"
            if state in self._accepting:
                markers += "*"
            label = f"{markers} {state}" if markers else f"  {state}"

            row = f"{label:<{state_width}}│"
            for event in sorted_events:
                target = self._transitions.get((state, event), "—")
                row += f" {target:<{event_width}} │"
            lines.append(row)

        return "\n".join(lines)

    def to_table(self) -> list[list[str]]:
        """Return the transition table as a list of rows.

        First row is the header: ["State", event1, event2, ...].
        Subsequent rows: [state_name, target1, target2, ...].
        Missing transitions are represented as "—".

        Returns:
            A list of string lists, suitable for formatting or export.
        """
        sorted_events = sorted(self._alphabet)
        sorted_states = sorted(self._states)

        rows: list[list[str]] = []
        rows.append(["State", *sorted_events])

        for state in sorted_states:
            row = [state]
            for event in sorted_events:
                target = self._transitions.get((state, event), "—")
                row.append(target)
            rows.append(row)

        return rows

    # === Equality and Representation ===

    def __repr__(self) -> str:
        """Return a readable representation of the DFA."""
        return (
            f"DFA(states={sorted(self._states)}, "
            f"alphabet={sorted(self._alphabet)}, "
            f"initial='{self._initial}', "
            f"accepting={sorted(self._accepting)}, "
            f"current='{self._current}')"
        )

File: code/packages/ruby/state_machine/lib/coding_adventures/state_machine/dfa.rb
# frozen_string_literal: true

require "set"

# ---------------------------------------------------------------------------
# Deterministic Finite Automaton (DFA) -- the workhorse of state machines.
# ---------------------------------------------------------------------------
#
# === What is a DFA? ===
#
# A DFA is the simplest kind of state machine. It has a fixed set of states,
# reads input symbols one at a time, and follows exactly one transition for
# each (state, input) pair. There is no ambiguity, no guessing, no backtracking.
#
# Formally, a DFA is a 5-tuple (Q, Sigma, delta, q0, F):
#
#     Q      = a finite set of states
#     Sigma  = a finite set of input symbols (the "alphabet")
#     delta  = a transition function: Q x Sigma -> Q
#     q0     = the initial state (q0 in Q)
#     F      = a set of accepting/final states (F is a subset of Q)
#
# === Why "deterministic"? ===
#
# "Deterministic" means there is exactly ONE next state for every (state, input)
# combination. Given the same starting state and the same input sequence, a DFA
# always follows the same path and reaches the same final state. This makes DFAs
# predictable, efficient, and easy to implement in hardware -- which is why they
# appear everywhere from CPU branch predictors to network protocol handlers.
#
# === Example: a turnstile ===
#
# A turnstile at a subway station has two states: locked and unlocked.
# Insert a coin -> it unlocks. Push the arm -> it locks.
#
#     States:      {locked, unlocked}
#     Alphabet:    {coin, push}
#     Transitions: (locked, coin)    -> unlocked
#                  (locked, push)    -> locked
#                  (unlocked, coin)  -> unlocked
#                  (unlocked, push)  -> locked
#     Initial:     locked
#     Accepting:   {unlocked}
#
# This DFA answers the question: "after this sequence of coin/push events,
# is the turnstile unlocked?"
#
# === Connection to existing code ===
#
# The 2-bit branch predictor in the branch-predictor package (D02) is a DFA:
#
#     States:      {SNT, WNT, WT, ST}  (strongly/weakly not-taken/taken)
#     Alphabet:    {taken, not_taken}
#     Transitions: defined by the saturating counter logic
#     Initial:     WNT
#     Accepting:   {WT, ST}  (states that predict "taken")
#
# The CPU pipeline (D04) is a linear DFA: FETCH -> DECODE -> EXECUTE -> repeat.
# The lexer's character dispatch is an implicit DFA where character classes
# determine transitions.
#
# === Ruby Implementation Notes ===
#
# - States and events are strings (just like the Python version).
# - Transitions are stored as a Hash with [state, event] array keys.
# - We use Ruby's Set class for states, alphabet, and accepting sets.
# - Actions are callable objects (Procs, lambdas, or anything responding to #call).
# - All validation happens eagerly in initialize ("fail fast" principle).
# ---------------------------------------------------------------------------

module CodingAdventures
  module StateMachine
    class DFA
      # The finite set of states (frozen Set of strings).
      attr_reader :states

      # The finite set of input symbols (frozen Set of strings).
      attr_reader :alphabet

      # The initial state (string).
      attr_reader :initial

      # The set of accepting/final states (frozen Set of strings).
      attr_reader :accepting

      # The state the machine is currently in (string).
      attr_reader :current_state

      # === Construction ===
      #
      # We validate all inputs eagerly in initialize so that errors are caught
      # at definition time, not at runtime when the machine processes its
      # first input. This is the "fail fast" principle.
      #
      # @param states [Set<String>] The finite set of states. Must be non-empty.
      # @param alphabet [Set<String>] The finite set of input symbols. Must be non-empty.
      # @param transitions [Hash<Array(String, String), String>] Mapping from
      #   [state, event] to target state. Every target must be in +states+.
      # @param initial [String] The starting state. Must be in +states+.
      # @param accepting [Set<String>] The set of accepting/final states.
      # @param actions [Hash<Array(String, String), #call>, nil] Optional mapping
      #   from [state, event] to a callable that fires when that transition occurs.
      # @raise [ArgumentError] If any validation check fails.
      def initialize(states:, alphabet:, transitions:, initial:, accepting:, actions: nil)
        # --- Validate states ---
        raise ArgumentError, "States set must be non-empty" if states.empty?

        # --- Validate initial state ---
        unless states.include?(initial)
          raise ArgumentError,
            "Initial state '#{initial}' is not in the states set #{states.to_a.sort}"
        end

        # --- Validate accepting states ---
        invalid_accepting = accepting - states
        unless invalid_accepting.empty?
          raise ArgumentError,
            "Accepting states #{invalid_accepting.to_a.sort} are not in " \
            "the states set #{states.to_a.sort}"
        end

        # --- Validate transitions ---
        #
        # Every transition must go FROM a known state ON a known event
        # TO a known state. We check all three.
        transitions.each do |(source, event), target|
          unless states.include?(source)
            raise ArgumentError,
              "Transition source '#{source}' is not in the states set"
          end
          unless alphabet.include?(event)
            raise ArgumentError,
              "Transition event '#{event}' is not in the alphabet #{alphabet.to_a.sort}"
          end
          unless states.include?(target)
            raise ArgumentError,
              "Transition target '#{target}' (from " \
              "(#{source}, #{event})) is not in the states set"
          end
        end

        # --- Validate actions ---
        if actions
          actions.each_key do |(source, event)|
            unless transitions.key?([source, event])
              raise ArgumentError,
                "Action defined for (#{source}, #{event}) but no " \
                "transition exists for that pair"
            end
          end
        end

        # --- Store the 5-tuple + extras ---
        @states = states.to_set.freeze
        @alphabet = alphabet.to_set.freeze
        @transitions = transitions.dup.freeze
        @initial = initial.freeze
        @accepting = accepting.to_set.freeze
        @actions = (actions || {}).dup.freeze

        # --- Mutable execution state ---
        @current_state = initial
        @trace = []
      end

      # The transition function as a hash. Returns a copy so callers cannot
      # modify the internal transitions.
      #
      # @return [Hash<Array(String, String), String>]
      def transitions
        @transitions.dup
      end

      # The execution trace -- a list of all TransitionRecords taken so far.
      # Returns a copy so callers cannot modify the internal trace.
      #
      # @return [Array<TransitionRecord>]
      def trace
        @trace.dup
      end

      # === Processing ===

      # Process a single input event and return the new state.
      #
      # Looks up the transition for (current_state, event), moves to the
      # target state, executes the action (if defined), logs a
      # TransitionRecord, and returns the new current state.
      #
      # How it works step by step:
      #
      #   1. Check that the event is in the alphabet. If not, raise.
      #   2. Look up [current_state, event] in the transition table.
      #   3. If no transition exists, raise (this is a "stuck" DFA).
      #   4. If an action is registered for this transition, call it
      #      with (source, event, target).
      #   5. Append a TransitionRecord to the trace.
      #   6. Update current_state to the target.
      #   7. Return the new current_state.
      #
      # @param event [String] An input symbol from the alphabet.
      # @return [String] The new current state after the transition.
      # @raise [ArgumentError] If the event is not in the alphabet, or if no
      #   transition is defined for (current_state, event).
      def process(event)
        # Validate the event
        unless @alphabet.include?(event)
          raise ArgumentError,
            "Event '#{event}' is not in the alphabet #{@alphabet.to_a.sort}"
        end

        # Look up the transition
        key = [@current_state, event]
        unless @transitions.key?(key)
          raise ArgumentError,
            "No transition defined for (state='#{@current_state}', " \
            "event='#{event}')"
        end

        target = @transitions[key]

        # Execute the action if one exists
        action_name = nil
        if @actions.key?(key)
          action = @actions[key]
          action.call(@current_state, event, target)
          # Try to get a meaningful name for the action
          action_name = if action.respond_to?(:name) && !action.name.empty?
            action.name
          else
            action.to_s
          end
        end

        # Log the transition
        record = TransitionRecord.new(
          @current_state,
          event,
          target,
          action_name
        )
        @trace << record

        # Move to the new state
        @current_state = target
        target
      end

      # Process a sequence of inputs and return the trace.
      #
      # Each input is processed in order. The full trace of transitions
      # is returned. The machine's state is updated after each input.
      #
      # @param events [Array<String>] A list of input symbols.
      # @return [Array<TransitionRecord>] One record per input processed.
      def process_sequence(events)
        trace_start = @trace.length
        events.each { |event| process(event) }
        @trace[trace_start..]
      end

      # Check if the machine accepts the input sequence.
      #
      # Processes the entire sequence and returns true if the machine
      # ends in an accepting state.
      #
      # IMPORTANT: This method does NOT modify the machine's current state
      # or trace. It runs on a fresh copy starting from the initial state.
      #
      # The algorithm is simple:
      #   1. Start at the initial state (ignoring current_state).
      #   2. For each event, look up the transition.
      #   3. If no transition exists, return false immediately (graceful reject).
      #   4. After all events, return true if the final state is accepting.
      #
      # @param events [Array<String>] A list of input symbols.
      # @return [Boolean] True if the machine accepts the sequence.
      # @raise [ArgumentError] If any event is not in the alphabet.
      def accepts(events)
        # Run on a copy so we don't modify this machine's state
        state = @initial
        events.each do |event|
          unless @alphabet.include?(event)
            raise ArgumentError,
              "Event '#{event}' is not in the alphabet #{@alphabet.to_a.sort}"
          end
          key = [state, event]
          return false unless @transitions.key?(key)
          state = @transitions[key]
        end
        @accepting.include?(state)
      end

      # Reset the machine to its initial state and clear the trace.
      #
      # After reset, the machine is in the same state as when it was
      # first constructed -- as if no inputs had ever been processed.
      def reset
        @current_state = @initial
        @trace = []
      end

      # === Introspection ===
      #
      # These methods analyze the structure of the DFA itself, not its
      # execution. They answer questions like "is the DFA well-formed?"
      # and "which states can actually be reached?"

      # Return the set of states reachable from the initial state.
      #
      # Uses breadth-first search over the transition graph. A state is
      # reachable if there exists any sequence of inputs that leads from
      # the initial state to that state.
      #
      # States that are defined but not reachable are "dead weight" --
      # they can never be entered and can be safely removed during
      # minimization.
      #
      # BFS algorithm:
      #   1. Start with a queue containing just the initial state.
      #   2. Pop a state from the queue.
      #   3. For each transition FROM that state, add the target to the
      #      queue if we haven't visited it yet.
      #   4. Repeat until the queue is empty.
      #   5. Return all visited states.
      #
      # @return [Set<String>] The set of reachable state names.
      def reachable_states
        visited = Set.new
        queue = [@initial]

        until queue.empty?
          state = queue.shift
          next if visited.include?(state)
          visited.add(state)

          # Find all states reachable from this one via any input
          @transitions.each do |(source, _event), target|
            if source == state && !visited.include?(target)
              queue << target
            end
          end
        end

        visited.freeze
      end

      # Check if a transition is defined for every (state, input) pair.
      #
      # A complete DFA never gets "stuck" -- every state handles every
      # input. Textbook DFAs are usually complete (missing transitions
      # go to an explicit "dead" or "trap" state). Practical DFAs often
      # omit transitions to save space, treating missing transitions as
      # errors.
      #
      # @return [Boolean] True if every (state, event) pair has a defined transition.
      def complete?
        @states.each do |state|
          @alphabet.each do |event|
            return false unless @transitions.key?([state, event])
          end
        end
        true
      end

      # Check for common issues and return a list of warnings.
      #
      # Checks performed:
      # - Unreachable states (defined but never entered)
      # - Missing transitions (incomplete DFA)
      # - Accepting states that are unreachable
      #
      # @return [Array<String>] Warning messages. Empty if no issues found.
      def validate
        warnings = []

        # Check for unreachable states
        reachable = reachable_states
        unreachable = @states - reachable
        unless unreachable.empty?
          warnings << "Unreachable states: #{unreachable.to_a.sort}"
        end

        # Check for unreachable accepting states
        unreachable_accepting = @accepting - reachable
        unless unreachable_accepting.empty?
          warnings << "Unreachable accepting states: #{unreachable_accepting.to_a.sort}"
        end

        # Check for missing transitions
        missing = []
        @states.to_a.sort.each do |state|
          @alphabet.to_a.sort.each do |event|
            unless @transitions.key?([state, event])
              missing << "(#{state}, #{event})"
            end
          end
        end
        unless missing.empty?
          warnings << "Missing transitions: #{missing.join(", ")}"
        end

        warnings
      end

      # === Visualization ===

      # Return a Graphviz DOT representation of this DFA.
      #
      # Accepting states are drawn as double circles (doublecircle shape).
      # The initial state has an invisible node pointing to it (the
      # standard convention for marking the start state in automata
      # diagrams).
      #
      # The output can be rendered with:
      #     dot -Tpng machine.dot -o machine.png
      #
      # @return [String] A string in DOT format.
      def to_dot
        lines = []
        lines << "digraph DFA {"
        lines << "    rankdir=LR;"
        lines << ""

        # Invisible start node pointing to initial state
        lines << "    __start [shape=point, width=0.2];"
        lines << "    __start -> \"#{@initial}\";"
        lines << ""

        # Accepting states get double circles
        @states.to_a.sort.each do |state|
          shape = @accepting.include?(state) ? "doublecircle" : "circle"
          lines << "    \"#{state}\" [shape=#{shape}];"
        end
        lines << ""

        # Transitions as labeled edges
        # Group transitions with same source and target to combine labels
        edge_labels = {}
        @transitions.sort.each do |(source, event), target|
          key = [source, target]
          edge_labels[key] ||= []
          edge_labels[key] << event
        end

        edge_labels.sort.each do |(source, target), labels|
          label = labels.sort.join(", ")
          lines << "    \"#{source}\" -> \"#{target}\" [label=\"#{label}\"];"
        end

        lines << "}"
        lines.join("\n")
      end

      # Return an ASCII transition table.
      #
      # Example output for the turnstile:
      #
      #           | coin     | push
      #   --------+----------+----------
      #   locked  | unlocked | locked
      #   unlocked| unlocked | locked
      #
      # Accepting states are marked with (*). The initial state is
      # marked with (>).
      #
      # @return [String] A formatted ASCII table string.
      def to_ascii
        sorted_events = @alphabet.to_a.sort
        sorted_states = @states.to_a.sort

        # Calculate column widths
        state_width = sorted_states.map { |s| s.length + 4 }.max # +4 for markers
        event_width = [
          sorted_events.map(&:length).max || 0,
          sorted_states.flat_map { |s|
            sorted_events.map { |e| (@transitions[[s, e]] || "\u2014").length }
          }.max || 0
        ].max
        event_width = [event_width, 5].max # minimum column width

        # Header row
        header = " " * state_width + "\u2502"
        sorted_events.each do |event|
          header += " #{event.ljust(event_width)} \u2502"
        end
        lines = [header]

        # Separator
        sep = "\u2500" * state_width + "\u253C"
        sorted_events.each do
          sep += "\u2500" * (event_width + 2) + "\u253C"
        end
        sep = sep[0..-2] # remove trailing cross
        lines << sep

        # Data rows
        sorted_states.each do |state|
          markers = ""
          markers += ">" if state == @initial
          markers += "*" if @accepting.include?(state)
          label = markers.empty? ? "  #{state}" : "#{markers} #{state}"

          row = "#{label.ljust(state_width)}\u2502"
          sorted_events.each do |event|
            target = @transitions[[state, event]] || "\u2014"
            row += " #{target.ljust(event_width)} \u2502"
          end
          lines << row
        end

        lines.join("\n")
      end

      # Return the transition table as an array of rows.
      #
      # First row is the header: ["State", event1, event2, ...].
      # Subsequent rows: [state_name, target1, target2, ...].
      # Missing transitions are represented as a dash character.
      #
      # @return [Array<Array<String>>]
      def to_table
        sorted_events = @alphabet.to_a.sort
        sorted_states = @states.to_a.sort

        rows = []
        rows << ["State", *sorted_events]

        sorted_states.each do |state|
          row = [state]
          sorted_events.each do |event|
            target = @transitions[[state, event]] || "\u2014"
            row << target
          end
          rows << row
        end

        rows
      end

      # Return a readable representation of the DFA.
      #
      # Includes the key components: states, alphabet, initial state,
      # accepting states, and current state. Useful for debugging and
      # REPL exploration.
      #
      # @return [String]
      def inspect
        "DFA(states=#{@states.to_a.sort}, " \
          "alphabet=#{@alphabet.to_a.sort}, " \
          "initial='#{@initial}', " \
          "accepting=#{@accepting.to_a.sort}, " \
          "current='#{@current_state}')"
      end

      alias_method :to_s, :inspect
    end
  end
end
